File: src/test/java/spring/graphql/modules/category/CategoryControllerTest.java
package spring.graphql.modules.category;

import static java.util.Optional.of;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.nextspring.modules.category.CategoryEntity;
import com.nextspring.modules.category.CategoryRepository;

@ExtendWith(MockitoExtension.class)
class CategoryControllerTest {

	@Mock
	private CategoryRepository categoryRepository;

	@Test
	void addCategory() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "name");

		when(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity);

		CategoryEntity result = categoryRepository.save(categoryEntity);

		assertEquals(categoryEntity, result);
		assertEquals(categoryEntity.getName(), result.getName());
	}

	@Test
	void categories() {
		List<CategoryEntity> categoryEntities = List.of(
				new CategoryEntity(null, "name1"),
				new CategoryEntity(null, "name2"),
				new CategoryEntity(null, "name3"));

		when(categoryRepository.findAll()).thenReturn(categoryEntities);

		List<CategoryEntity> result = categoryRepository.findAll();

		assertEquals(categoryEntities, result);
		assertEquals(categoryEntities.size(), result.size());
	}

	@Test
	void category() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "name");

		when(categoryRepository.findById(categoryEntity.getId())).thenReturn(of(categoryEntity));

		CategoryEntity result = categoryRepository.findById(categoryEntity.getId()).orElseThrow();

		assertEquals(categoryEntity, result);
		assertEquals(categoryEntity.getName(), result.getName());
	}

	@Test
	void updateCategory() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "name");

		when(categoryRepository.findById(categoryEntity.getId())).thenReturn(of(categoryEntity));

		CategoryEntity result = categoryRepository.findById(categoryEntity.getId()).orElseThrow();

		assertEquals(categoryEntity, result);
		assertEquals(categoryEntity.getName(), result.getName());
	}

	@Test
	void deleteCategory() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "name");

		when(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity);

		CategoryEntity savedCategory = categoryRepository.save(categoryEntity);
		assertNotNull(savedCategory);
		assertEquals(categoryEntity, savedCategory);

		doNothing().when(categoryRepository).delete(categoryEntity);
		categoryRepository.delete(categoryEntity);

		verify(categoryRepository, times(1)).delete(categoryEntity);
	}
}

File: src/test/kotlin/spring/graphql/modules/category/CategoryControllerTest.kt
package spring.graphql.modules.category

import com.nextspring.modules.category.CategoryEntity
import com.nextspring.modules.category.CategoryRepository
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*

@ExtendWith(MockitoExtension::class)
internal class CategoryControllerTest {
    @Mock
    private val categoryRepository: CategoryRepository = Mockito.mock(CategoryRepository::class.java)

    @Test
    fun addCategory() {
        val categoryEntity = CategoryEntity(null, "name")

        Mockito.`when`(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity)

        val result = categoryRepository.save(categoryEntity)

        Assertions.assertEquals(categoryEntity, result)
        Assertions.assertEquals(categoryEntity.name, result.name)
    }

    @Test
    fun categories() {
        val categoryEntities = listOf(
            CategoryEntity(null, "name1"),
            CategoryEntity(null, "name2"),
            CategoryEntity(null, "name3")
        )

        Mockito.`when`(categoryRepository.findAll()).thenReturn(categoryEntities)

        val result = categoryRepository.findAll()

        Assertions.assertEquals(categoryEntities, result)
        Assertions.assertEquals(categoryEntities.size, result.size)
    }

    @Test
    fun category() {
        val categoryEntity = CategoryEntity(null, "name")

        Mockito.`when`(categoryRepository.findById(categoryEntity.id!!)).thenReturn(Optional.of(categoryEntity))

        val result = categoryRepository.findById(categoryEntity.id!!).orElseThrow()!!

        Assertions.assertEquals(categoryEntity, result)
        Assertions.assertEquals(categoryEntity.name, result.name)
    }

    @Test
    fun updateCategory() {
        val categoryEntity = CategoryEntity(null, "name")

        Mockito.`when`(categoryRepository.findById(categoryEntity.id!!)).thenReturn(Optional.of(categoryEntity))

        val result = categoryRepository.findById(categoryEntity.id!!).orElseThrow()!!

        Assertions.assertEquals(categoryEntity, result)
        Assertions.assertEquals(categoryEntity.name, result.name)
    }

    @Test
    fun deleteCategory() {
        val categoryEntity = CategoryEntity(null, "name")

        Mockito.`when`(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity)

        val savedCategory = categoryRepository.save(categoryEntity)
        Assertions.assertNotNull(savedCategory)
        Assertions.assertEquals(categoryEntity, savedCategory)

        Mockito.doNothing().`when`(categoryRepository).delete(categoryEntity)
        categoryRepository.delete(categoryEntity)

        Mockito.verify(categoryRepository, Mockito.times(1)).delete(categoryEntity)
    }
}